### PR TITLE
Publish to Maven

### PR DESCRIPTION
**Pull request by jcnoriega**
Added job to publish to Maven. 
Also added complete release workflow.

Screen Shot 2021-08-20 at 13 06 58

### CODE REVIEW

File: Card/build.gradle
Comment (sarahkoop) [inline Card/build.gradle]: Let's pull all these into the top level gradle (I think `versionName` can be the same for all modules)

File: build.gradle
Comment (sarahkoop) [inline build.gradle]: Did we need to downgrade the build tools version for publishing?

File: gradle.properties
Comment (sarahkoop) [inline gradle.properties]: Should these be using the `git@github` urls instead?

File: gradle.properties
Comment (sarahkoop) [inline gradle.properties]: I think these need to be specific to each module not project level

File: gradle/gradle-publish.gradle
Comment (sarahkoop) [inline gradle/gradle-publish.gradle:36]: Is email required? If we aren't defining it and it's not required lets remove

File: gradle/gradle-publish.gradle
Comment (sarahkoop) [inline gradle/gradle-publish.gradle]: Do we want this `println`? Is it printing to the GH actions console?

File: gradle/gradle-publish.gradle
Comment (sarahkoop) [inline gradle/gradle-publish.gradle]: Do these variables need to be defined separately or can we just access `variant.name` and `project.version` directly?

File: gradle/gradle-publish.gradle
Comment (sarahkoop) [inline gradle/gradle-publish.gradle]: Is `project.name` the module name here? So card and core can have their own artifact IDs?

File: gradle/gradle-publish.gradle
Comment (jcnoriega) [reply gradle/gradle-publish.gradle]: I added this to debug, i'll remove it

File: Card/build.gradle
Comment (jcnoriega) [reply Card/build.gradle]: @sarahkoop do you mean that every module is going to have the same `versionName`?

File: gradle/gradle-publish.gradle
Comment (jcnoriega) [reply gradle/gradle-publish.gradle]: yes, `project.name` is the module name. So we upload  `com.paypal.android.core` and `com.paypal.android.card` for example

File: Card/build.gradle
Comment (sarahkoop) [reply Card/build.gradle]: Yes in `braintree_android` this field is set at [top level](https://github.com/braintree/braintree_android/blob/master/build.gradle#L35) and corresponds to the version to be released like `1.0.0-SNAPSHOT` or `1.1.0` or 1.0.0-beta`. All modules should be on the same version I believe.

Comment (sshropshire) [reply Card/build.gradle]: @jcnoriega Yeah historically it's been a bit of a compromise. Separate version numbers complicate git tagging and releases in general. We went with a monolith versioning scheme. We mainly use semantic versioning to communicate new features / breaking changes to merchants.

File: .github/workflows/build.yml
Comment (sshropshire) [inline .github/workflows/build.yml:26]: random q: where do we go to manage secrets in GH actions?

Comment (jcnoriega) [reply .github/workflows/build.yml:26]: you need to have admin access to the repo, then go to Settings > Secrets.

File: gradle.properties
Comment (jcnoriega) [reply gradle.properties]: i copied this from our checkout sdk, but then after reading a bit, it turns out that that is the way maven requires it: https://maven.apache.org/scm/git.html

File: .github/workflows/build.yml
Comment (sshropshire) [reply .github/workflows/build.yml:26]: 👍  TIL

File: Card/build.gradle
Comment (sarahkoop) [inline Card/build.gradle]: Can we extract these numbers in Card and Core into the shared gradle?

File: build.gradle
Comment (sarahkoop) [inline build.gradle]: Nitpick but I think we are going with all double quotes for standardization

File: gradle.properties
Comment (sarahkoop) [reply gradle.properties]: Does Sonatype require this too? In BT Android we use git@github. Not sure which is better here

File: gradle.properties
Comment (sarahkoop) [inline gradle.properties:35]: Are these going to be addressed in a separate PR? Let's define them now if we can

File: gradle/gradle-publish.gradle
Comment (sarahkoop) [inline gradle/gradle-publish.gradle]: We might want some additional logic here for the artifact naming. Do we want dashes rather than all lower case for modules with two words (i.e. `core-payments` rather than `corepayments`)?

File: gradle/gradle-publish.gradle
Comment (sarahkoop) [inline gradle/gradle-publish.gradle]: Not sure exactly the kotlin equivalent of javadocs but we are missing publishing those here - Should we add a TODO somewhere so we don't forget that before release?

File: .github/workflows/build.yml
Comment (tdchow) [inline .github/workflows/build.yml:7]: Do we need to have these signing steps in our build workflow?

File: .github/workflows/publish_to_maven.yml
Comment (tdchow) [inline .github/workflows/publish_to_maven.yml]: Currently, these publish steps will run in sequence. Did we want to create separate jobs for each that will run in parallel?

File: Core/build.gradle
Comment (tdchow) [inline Core/build.gradle]: Is there a reason why we bumped versionCode to 2?

File: gradle/gradle-publish.gradle
Comment (tdchow) [reply gradle/gradle-publish.gradle]: I vote for using dashes in the artifactIDs.

File: Core/build.gradle
Comment (jcnoriega) [reply Core/build.gradle]: Oops, forgot about this, was doing some testing

File: .github/workflows/build.yml
Comment (jcnoriega) [reply .github/workflows/build.yml:7]: yes, because the assemble task tries to sign the artifacts, so i need to include the signing keys in the build

File: gradle/gradle-publish.gradle
Comment (jcnoriega) [reply gradle/gradle-publish.gradle]: I added `toLowerCase` to transform `Card` and `Code` into `card`, `code`. If a module has two words, i would add the dash on the module name.

File: gradle/gradle-publish.gradle
Comment (jcnoriega) [reply gradle/gradle-publish.gradle]: im good with adding a TODO, im not really sure how to generate those. I dont want to just paste something from `stackoverflow`. Maybe we can also add a ticket and sync later on it. Thoughts?

File: gradle.properties
Comment (jcnoriega) [reply gradle.properties:35]: This is the one we have in our checkout repo: https://developer.paypal.com/docs/limited-release/native-checkout/license/
Im not sure if it applies though. Perhaps we can add TODO and create a ticket for it until we are sure?

File: gradle/gradle-publish.gradle
Comment (sarahkoop) [reply gradle/gradle-publish.gradle]: I'd opt to pull the artifact names into their own gradle files along with the POM name and descriptions (since those should be module specific). Dashes in module names doesn't feel like Kotlin/Java convention to me. Thoughts?

File: gradle/gradle-publish.gradle
Comment (sarahkoop) [reply gradle/gradle-publish.gradle]: Yeah let's just create a JIRA card for that

File: gradle/gradle-publish.gradle
Comment (jcnoriega) [reply gradle/gradle-publish.gradle]: @sarahkoop agreed

File: gradle.properties
Comment (jcnoriega) [reply gradle.properties]: Me neither, i dont mind putting git@github

File: .github/workflows/publish_to_maven.yml
Comment (sarahkoop) [inline .github/workflows/publish_to_maven.yml]:
```suggestion
on: [workflow_dispatch]
```
Don't want to release on every pull request

File: .github/workflows/release.yml
Comment (sarahkoop) [inline .github/workflows/release.yml]:
```suggestion
on: [workflow_dispatch]
```

File: .github/workflows/publish_to_maven.yml
Comment (jcnoriega) [reply .github/workflows/publish_to_maven.yml]: @sarahkoop I had to add `pull_request` because of a bug in github actions. If i only add `workflow_dispatch` it wont appear, unless that file is already on `main` branch. I was going to delete them once this PR is approved

Comment (sarahkoop) [reply .github/workflows/publish_to_maven.yml]: Maybe I'm confused - it looks like the release github workflow is executing successfully, what is preventing these artifacts from being published automatically currently?

Comment (jcnoriega) [reply .github/workflows/publish_to_maven.yml]: its like this: github actions has a bug. If your `.yml` is not on your default branch (in our case `main` branch) and your workflow ONLY has `workflow_dispatch`, you wont be able to see it in the github UI. So, in order for me to test it, i had to add `pull_request`, and only then the workflow appeared in the github UI. Before merging it to main, i'm going to delete `pull_request` tag, but until i get all the approvals, i need to leave it there so that i can test the workflow. Does that make sense?

**Review by sshropshire — APPROVED** [review]

**Review by sarahkoop — APPROVED** [review]
Looks good to me - pending removing `pull_request` as the GH workflow trigger, and adding a step to close and release the repository in the future

**Pull request merged**
